liquid/validation.go

```diff
@@ -0,0 +1,316 @@
+	// and forget to update this function accordingly
+	var isFlat bool
+	switch topology {
+	case FlatTopology:
+		isFlat = true
+	case AZAwareTopology, AZSeparatedTopology:
+		isFlat = false
+	default:
+		if topology.IsValid() {
+			return fmt.Errorf("%s[%q] has topology %q, but validatePerAZAgainstTopology() has not been updated to understand this value",
+				path, name, topology)
+		} else {
+			// it should not be possible to reach this point,
+			// callers should already have rejected invalid topology values
+			panic(fmt.Sprintf("unreachable: topology = %q", topology))
+		}
+	}
+
+	valid := true // until proven otherwise
+	var expectedAZs []AvailabilityZone
+	if isFlat {
+		// Flatopology requires "any" and allows nothing else
+		if len(perAZ) != 1 {
+			valid = false
+		}
+		for az := range perAZ {
+			if az != AvailabilityZoneAny {
+				valid = false
+			}
+		}
+		expectedAZs = []AvailabilityZone{AvailabilityZoneAny}
+	} else {
+		// other topologies require each AZ from `allAZs` to be present, and then optionally allow "unknown", but nothing else
+		for az := range perAZ {
+			if az != AvailabilityZoneUnknown && !slices.Contains(allAZs, az) {
+				valid = false
+			}
+		}
+		for _, az := range allAZs {
+			if !hasKey(perAZ, az) {
+				valid = false
+			}
+		}
+		expectedAZs = allAZs
+	}
+
+	if !valid {
+		return fmt.Errorf("%s[%q].PerAZ has entries for %#v, which is invalid for topology %q (expected entries for %#v)",
+			path, name, slices.Sorted(maps.Keys(perAZ)), topology, expectedAZs)
+	}
+	return nil
+}
+
+func validateQuotaAgainstTopology(report *ResourceUsageReport, hasQuota bool, topology Topology, name ResourceName, allAZs []AvailabilityZone) error {
+	// report.Quota shall be null if and only if the resource is declared with "HasQuota = false" or with AZSeparatedTopology
+	if report.Quota == nil && hasQuota && topology != AZSeparatedTopology {
+		return fmt.Errorf(".Resources[%q] has no quota reported on resource level, which is invalid for HasQuota = true and topology %q", name, topology)
+	}
+	if report.Quota != nil {
+		if !hasQuota {
+			return fmt.Errorf(".Resources[%q] has quota reported on resource level, which is invalid for HasQuota = false", name)
+		}
+		if topology == AZSeparatedTopology {
+			return fmt.Errorf(".Resources[%q] has quota reported on resource level, which is invalid for topology %q", name, topology)
+		}
+	}
+
+	var allAZsWithoutQuota []string
+	for _, az := range allAZs {
+		azReport, exists := report.PerAZ[az]
+		if !exists || azReport.Quota == nil {
+			allAZsWithoutQuota = append(allAZsWithoutQuota, string(az))
+			continue
+		}
+		// azReport.Quota shall be non-null if and only if the resource is declared with AZSeparatedTopology
+		if azReport.Quota != nil && hasQuota && topology != AZSeparatedTopology {
+			return fmt.Errorf(".Resources[%q] has quota reported on AZ level, which is invalid for topology %q", name, topology)
+		}
```

Comment:
There should also be a check for `hasQuota == false` here, same as for the resource-level quota above.